cmake/shields.cmake

```diff
@@ -1,18 +1,138 @@
-if(CMAKE_SCRIPT_MODE_FILE AND NOT CMAKE_PARENT_LIST_FILE)
-  # This file was invoked as a script directly with -P:
-  # cmake -P shields.cmake
-  #
-  # Unlike boards.cmake, this takes no OUTPUT_FILE option, but
-  # SHIELD_LIST_SPACE_SEPARATED is required.
-  list(SORT SHIELD_LIST)
-  foreach(shield ${SHIELD_LIST})
-    message("${shield}")
+# SPDX-License-Identifier: Apache-2.0
+#
+# Copyright (c) 2021, Nordic Semiconductor ASA
+
+# Validate shields and setup shields target.
+#
+# This module will validate the SHIELD argument.
+#
+# If a shield implementation is not found for one of the specified shields an
+# error will be raised and list of valid shields will be printed.
+#
+# Outcome:
+# The following variables will be defined when this module completes:
+# - shield_conf_files: List of shield specific Kconfig fragments
+# - shield_dts_files : List of shield specific devicetree files
+# - shield_dts_fixups: List of shield specific devicetree fixups
+#
+# The following targets will be defined when this module completes:
+# - shields: when invoked a list of valid shields will be printed
+#
+# Variable dependencies:
+# - BOARD_ROOT: CMake list of board roots containing board implementations
+#
+# Module dependencies:
+# - extension module.
```

Comment:
>  In general, I'm not sure the module dependencies comments are all correct in this PR; could you please double check those?

Maybe later patches in the series which just use `include` to source dependencies mean that maintaining these comments explicitly is not required.

Reply:
> - I don't understand what extension module. means. This is a general comment -- I think it would be clearer to say exactly the file names that this depends on, like extensions.cmake instead of this.

Except that we now include them as modules and not the files directly.
The main intention with adding this text in top of CMake modules files is to prepare to generate rst docs to be automatically included in our doc-set, that way users can get an overview of Zephyr CMake modules and what they rely on.

I agree that with this new approach, `extensions.cmake` may not be the best chosen name.
But I would like to keep such discussion outside this PR.

I can update the comment to:
```
# CMake module dependencies: 
# - Zephyr CMake extension module. 
```

> Isn't boards.cmake a dependency due to the BOARD_MESSAGE mutation?

This has been removed now.

> In general, I'm not sure the module dependencies comments are all correct in this PR; could you please double check those?

Generally speaking, variables used inside a CMake module should behave sane if not-defined (empty), or it should have a sane default value.
This ensures that users wanting to use Zephyr package with a specific module can decide whether to have such variables not-defined, initialized before calling `find_package(Zephyr)`, or by also loading another module, for example `zephyr_module.cmake`.

Only when a missing module will cause errors, such as missing functions (example `extensions.cmake`), or external programs (example `python`), the providing module must be loaded.
See also here: https://github.com/zephyrproject-rtos/zephyr/pull/41301#discussion_r777444109 and https://github.com/zephyrproject-rtos/zephyr/pull/41301#discussion_r777959858